### PR TITLE
Poly canonicalization

### PR DESCRIPTION
**Pull request by j2kun**
Adds simple canonicalization rules to the polynomial dialect. Mainly to get the boilerplate incorporated before more substantial canonicalization patterns are added.

### CODE REVIEW

File: mlir/lib/Dialect/Polynomial/IR/PolynomialOps.cpp [hunk 1]
```diff
@@ -183,3 +185,26 @@ LogicalResult INTTOp::verify() {
...
+
+void NTTOp::getCanonicalizationPatterns(RewritePatternSet &results,
+                                        MLIRContext *context) {
+  populateWithGenerated(results);
+}
+
+void INTTOp::getCanonicalizationPatterns(RewritePatternSet &results,
+                                         MLIRContext *context) {
+  populateWithGenerated(results);
+}
```
Comment (joker-eph) [inline mlir/lib/Dialect/Polynomial/IR/PolynomialOps.cpp:210]: I haven't looked at our DRR infra for a while, but aren't we gonna populate these patterns many times that way?

@jpienaar may know...

Comment (j2kun) [reply mlir/lib/Dialect/Polynomial/IR/PolynomialOps.cpp:210]: Populated with the specific pattern names in ab63790e45f7bdcd2b2a8b0a4db05bfe39686679

**Review by joker-eph — APPROVED** [review]

File: mlir/lib/Dialect/Polynomial/IR/CMakeLists.txt [hunk 2]
```diff
@@ -1,3 +1,7 @@
+set(LLVM_TARGET_DEFINITIONS PolynomialCanonicalization.td)
+mlir_tablegen(PolynomialCanonicalization.inc -gen-rewriters)
+add_public_tablegen_target(MLIRPolynomialCanonicalizationIncGen)
```
Comment (joker-eph) [inline mlir/lib/Dialect/Polynomial/IR/CMakeLists.txt:3]: Don't you need to depend on MLIRPolynomialCanonicalizationIncGen somewhere?

Comment (ftynse) [reply mlir/lib/Dialect/Polynomial/IR/CMakeLists.txt:3]: +1. Bazel has such a dependency.

**Review by ftynse — APPROVED** [review]
LGTM with Mehdi's comment addressed.

File: mlir/lib/Dialect/Polynomial/IR/CMakeLists.txt (hunk 2 above)
Comment (j2kun) [reply mlir/lib/Dialect/Polynomial/IR/CMakeLists.txt:3]: In principle I agree (added in 0e25a328d9d05c080bef4e81df276b4f5a2b3a77).

In practice I am baffled by how CI and my local tests are passing without this dependency added. Excuse my build system naïveté, but maybe someone could help me understand why? Without that, I can't even be 100% sure I added the dependency in the right place.

Comment (joker-eph) [reply mlir/lib/Dialect/Polynomial/IR/CMakeLists.txt:3]: Most likely it would be concurrency and luck.

**Comment by clementval** [comment]
This breaks some flang buildbots. https://lab.llvm.org/buildbot/#/builders/268/builds/14288

**Comment by j2kun** [comment]
Fix in https://github.com/llvm/llvm-project/pull/92585

**Pull request merged**